src/saltext/vmware/states/roles.py

```diff
@@ -0,0 +1,161 @@
+    )
+    ret = {"name": name, "changes": {}, "result": None, "comment": ""}
+
+    # # Clone config input to a Map.
+    # # Can be used to transform input to internal objects and do validation if needed
+    new_config = {}
+    for role_config in config:
+        # Create full representation of the object, default or empty values
+        new_role_config = {"privileges": {}}
+        # Transform / Validate input vs object, e.g. constraints section
+        if "groups" in role_config:
+            for privilege in role_config["groups"]:
+                priv_group = privilege["group"]
+                new_role_config["privileges"][priv_group] = []
+                for item in privilege["privileges"]:
+                    new_role_config["privileges"][priv_group].append(item)
+        new_config[role_config["role"]] = new_role_config
+
+    log.debug("---------------NEW--------------")
+    log.debug(json.dumps(new_config, indent=2))
+    log.debug("---------------NEW--------------")
+
+    # Get all policies from vCenter, the objects are VMOMI objects
+    old_configs = roles_module.find(
+        role_name=None, service_instance=service_instance, profile=profile
+    )
+
+    # old_configs must have only the rules that are in the scope of interest (provided in argument config_input)
+    old_config = {}
+    for role in old_configs:
+        if role["role"] not in new_config.keys():
+            continue
+
+        old_config[role["role"]] = {"privileges": role["privileges"]}
+
+    log.debug("--------------OLD---------------")
+    log.debug(json.dumps(old_config, indent=2))
+    log.debug("--------------OLD---------------")
+
+    # Find rules changes
+    changes = []
+    diffs = drift.drift_report(old_config, new_config, diff_level=0)
+    diffs = json.loads(json.dumps(diffs))  # clone object
+    if diffs is not None:
+        ret["changes"] = diffs
+
+        log.debug("==============DRIFT===============")
+        log.debug(json.dumps(diffs, indent=2))
+        log.debug("==============DRIFT===============")
+
+        # add changes for process if not dry-run
+        for d_name in diffs:
+            new_policy = diffs[d_name]["new"]
+            changes.append({**{"role": d_name}, **new_policy})
+
+    # If it's not dry-run and has changes, then apply changes
+    if not __opts__["test"] and changes:
+        success = True
+
+        log.debug("===============CHANGES==============")
+        log.debug(json.dumps(changes, indent=2))
+        log.debug("===============CHANGES==============")
+
+        comments = {}
+        for change in changes:
+            try:
+                # save/update rule
+                roles_module.save(change, service_instance, profile)
+                comments[change["role"]] = {
+                    "status": "SUCCESS",
+                    "message": f"Role '{change['role']}' has been changed successfully.",
+                }
+            except Exception as err:
```

Comment:
**suggestion** at the location of a call, exceptions should be as narrow as possible. Broadly catching `Exception` is a recipe for hiding bugs.

In my particular case, I'm trying to run

```
Do my things:                                                                                                                                                                                 
  vcenter_roles.config:
    - config:
      - role: Cool Guy
        groups:
          - group: SRM protection
            privileges:
              - Stop
              - Protect
              - Attacc
```

And it's blowing up with a `KeyError`.

Also, note that `comments` is expected to be a text field, and `changes` is a data field. For example:

```
----------
          ID: Do my things
    Function: vcenter_roles.config
      Result: True
     Comment: {'Cool Guy': {'status': 'SUCCESS', 'message': "Role 'Cool Guy' has been changed successfully."}}
     Started: 18:25:44.232885
    Duration: 3223.102 ms
     Changes:   
              ----------
              Cool Guy:
                  ----------
                  new:
                      ----------
                      privileges:
                          ----------
                          Content Library:
                              - Read storage
                              - Protect
                              - Attacc
                  old:
                      ----------
                      privileges:
                          ----------
                          Content Library:

Summary for local
------------
Succeeded: 1 (changed=1)
Failed:    0
------------
Total states run:     1
Total run time:   3.223 s
```

It gets worse when more bits are added.

```
     Comment: {'New Guy': {'status': 'SUCCESS', 'message': "Role 'New Guy' has been changed successfully."}, 'Cool Guy': {'status': 'SUCCESS', 'message': "Role 'Cool Guy' has been changed successfully."}}
```

A summary of what went on would be fine, e.g.

```
changed = 0
errors = 0
for change in changes:
    try:
        ....
        changed += 1
     except ...:
         errors += 1
         
comment = "{} changes{}".format(changes, "." if not errors else f", {errors} errors.")
```